File: sorting/mergesort.js
function merge(left, right) {
    let result = []
    let i = 0
    let j = 0
    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            result.push(left[i])
            i++
        } else if (right[j] < left[i]) {
            result.push(right[j])
            j++
        } else {
            result.push(left[i])
            result.push(right[j])
            i++
            j++
        }
    }

    result = result.concat(left.slice(i))
    result = result.concat(right.slice(j))

    return result
}

function mergeSort(arr) {
    if (arr.length < 2) {
        return arr
    }

    const mid = Math.floor(arr.length / 2)
    return merge(mergeSort(arr.slice(0, mid)), mergeSort(arr.slice(mid)))
}

module.exports = mergeSort

File: sorting/mergesort.py
def merge(left, right):
    result = []
    i = 0
    j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        elif left[i] > right[j]:
            result.append(right[j])
            j += 1
        else:
            result.append(left[i])
            result.append(right[j])
            i += 1
            j += 1
    
    result.extend(left[i:])
    result.extend(right[j:])

    return result


def merge_sort(arr: list) -> list:
    if len(arr) < 2:
        return arr

    mid = len(arr) // 2
    return merge(merge_sort(arr[:mid]), merge_sort(arr[mid:]))